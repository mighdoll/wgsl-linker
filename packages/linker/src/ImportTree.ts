export class ImportTree {
  constructor(public segments: PathSegment[]) {}
}

export type PathSegment = SimpleSegment | Wildcard | ImportTree | SegmentList;


export class SimpleSegment {
  constructor(
    public name: string,
    public as?: string
  ) {}
}

export class SegmentList {
  constructor(public list: PathSegment[]) {}
}

export class Wildcard {}

export const Wildcard_ = new Wildcard();