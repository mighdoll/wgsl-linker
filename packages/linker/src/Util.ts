export function multiKeySet<A, B, V>(
  m: Map<A, Map<B, V>>,
  a: A,
  b: B,
  v: V
): void {
  const bMap = m.get(a) || new Map();
  m.set(a, bMap);
  bMap.set(b, v);
}

const tokenRegex = /\b(\w+)\b/gi;
/** replace strings in a text according to a relacement map
 * replaced strings must be 'tokens', surrounded by spaces or punctuation
 */
export function replaceWords(
  text: string,
  replace: Record<string, string>
): string {
  return text.replaceAll(tokenRegex, (s) => (s in replace ? replace[s] : s));
}

/** return an array partitioned into possibly overlapping groups */
export function grouped<T>(a: T[], size: number, stride = size): T[][] {
  const groups = [];
  for (let i = 0; i < a.length; i += stride) {
    groups.push(a.slice(i, i + size));
  }
  return groups;
}

/** group an array into subarrays by a key function */
export function groupBy<T, K>(a: T[], key: (t: T) => K): Map<K, T[]> {
  const groups = new Map<K, T[]>();
  for (const t of a) {
    const k = key(t);
    const group = groups.get(k) || [];
    group.push(t);
    groups.set(k, group);
  }
  return groups;
}

/** partition an array into two parts by a discriminator function */
export function partition<T>(a: T[], partFn: (t: T) => boolean): [T[], T[]] {
  const yesPart: T[] = [];
  const noPart: T[] = [];
  for (const t of a) {
    partFn(t) ? yesPart.push(t) : noPart.push(t);
  }
  return [yesPart, noPart];
}


/** run an carrying function over every element in an array */
export function scan<T, U>(array: T[], fn: (a: T, b: U) => U, zero: U): U[] {
  const result = [zero];

  let current = zero;
  for (let i = 0; i < array.length; i++) {
    current = fn(array[i], current);
    result.push(current);
  }
  return result;
}