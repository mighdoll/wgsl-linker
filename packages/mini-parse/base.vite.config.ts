import { UserConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import dts from "vite-plugin-dts";
import { resolve } from "path";

export function baseViteConfig(): UserConfig {
  return {
    plugins: [
      tsconfigPaths(),
      dts(), // generate .d.ts files
    ],
    build: {
      lib: {
        name: "mini-parse",
        entry: [resolve(__dirname, "src/index.ts")],
      },
      minify: false,
      sourcemap: true,
    },
  };
}
