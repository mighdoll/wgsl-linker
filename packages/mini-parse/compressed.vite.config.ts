/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
// import { visualizer } from "rollup-plugin-visualizer";
import { baseViteConfig } from "./vite-base-config.js";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
