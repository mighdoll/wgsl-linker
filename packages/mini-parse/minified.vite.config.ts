/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
// import { visualizer } from "rollup-plugin-visualizer";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
