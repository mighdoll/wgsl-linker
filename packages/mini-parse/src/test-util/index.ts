export * from "./LogCatcher.js";
export * from "./TestParse.js";
