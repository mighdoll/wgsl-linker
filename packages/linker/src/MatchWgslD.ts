import { matchOneOf, tokenMatcher } from "mini-parse";

/** token matchers for wgsl with #directives */

export const eol = /\n/;
export const directive = /#[a-zA-Z_]\w*/;
export const notDirective = /[^#\n]+/;

const symbolSet =
  "& && -> @ / ! [ ] { } : , = == != > >= < << <= % - -- " + // '>>' elided for template parsing, e.g. vec2<vec2<u8>>
  ". + ++ | || ( ) ; * ~ ^ // /* */ += -= *= /= %= &= |= ^= >>= <<= <<";
const symbol = matchOneOf(symbolSet);

/** matching tokens at wgsl root level */
export const mainTokens = tokenMatcher(
  {
    directive,
    attr: /@[a-zA-Z_]\w*/,
    word: /[a-zA-Z_]\w*/, // LATER consider making this 'ident' per wgsl spec (incl. non-ascii)   word,
    digits: /(?:0x)?[\d.]+[iuf]?/, // LATER parse more wgsl number variants
    symbol,
    ws: /\s+/,
  },
  "main"
);

export const moduleTokens = tokenMatcher(
  {
    ws: /\s+/,
    moduleName: /[a-zA-Z_][\w./-]*/,
  },
  "moduleName"
);

/** matching tokens at the start of a '//' line comment that might contain #directives */
export const lineCommentTokens = tokenMatcher(
  {
    directive,
    ws: /[ \t]+/, // note ws must be before notDirective
    notDirective,
    eol,
  },
  "lineComment"
);

/** matching tokens while parsing directive parameters #export foo(param1, param2) */
export const argsTokens = tokenMatcher(
  {
    directive,
    relPath: /[.][/\w._-]+/,
    arg: /[\w._-]+/,
    symbol,
    ws: /[ \t]+/, // don't include \n, so we can find eol separately
    eol,
  },
  "argsTokens"
);
