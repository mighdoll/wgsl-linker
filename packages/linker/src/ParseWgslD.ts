import {
  anyNot,
  anyThrough,
  eof,
  fn,
  kind,
  matchingLexer,
  opt,
  or,
  Parser,
  ParserContext,
  ParserInit,
  preParse,
  repeat,
  req,
  seq,
  setTraceName,
  simpleParser,
  SrcMap,
  tracing,
  withSep,
} from "mini-parse";
import {
  AbstractElem,
  AliasElem,
  CallElem,
  FnElem,
  FnNameElem,
  GlobalDirectiveElem,
  StructElem,
  StructMemberElem,
  TypeNameElem,
  TypeRefElem,
  VarElem,
} from "./AbstractElems.js";
import { mainTokens } from "./MatchWgslD.js";
import { directive } from "./ParseDirective.js";
import {
  comment,
  makeElem,
  unknown,
  word,
  wordNumArgs,
} from "./ParseSupport.js";

/** parser that recognizes key parts of WGSL and also directives like #import */

// prettier gets confused if we leave the quoted parens inline so make consts for them here
const lParen = "(";
const rParen = ")";

export interface ParseState {
  ifStack: boolean[]; // stack used while processiing nested #if #else #endif directives
  params: Record<string, any>; // user provided params to templates, code gen and #if directives
}

const optAttributes = repeat(seq(kind(mainTokens.attr), opt(wordNumArgs)));
const possibleTypeRef = Symbol("typeRef");

const globalDirectiveOrAssert = seq(
  or("diagnostic", "enable", "requires", "const_assert"),
  req(anyThrough(";"))
).map((r) => {
  const e = makeElem<GlobalDirectiveElem>("globalDirective", r);
  r.app.state.push(e);
});

/** parse an identifier into a TypeNameElem */
export const typeNameDecl = req(word.named("name")).map((r) => {
  return makeElem<TypeNameElem>("typeName", r, ["name"]);
});

/** parse an identifier into a TypeNameElem */
export const fnNameDecl = req(word.named("name"), "missing fn name").map(
  (r) => {
    return makeElem<FnNameElem>("fnName", r, ["name"]);
  }
);

/** find possible references to user types (structs) in this possibly nested template */
export const template: Parser<any> = seq(
  "<",
  or(
    word.named(possibleTypeRef), // only the first element of the template can be a type
    fn(() => template)
  ),
  repeat(
    or(
      fn(() => template),
      anyNot(">") // we don't care about the rest of the template
    )
  ),
  req(">")
);

/** find possible references to user structs in this type specifier and any templates */
export const typeSpecifier: Parser<TypeRefElem[]> = seq(
  word.named(possibleTypeRef),
  opt(template)
).map((r) =>
  r.named[possibleTypeRef].map((name) => {
    const e = makeElem<TypeRefElem>("typeRef", r);
    e.name = name;
    return e;
  })
);

export const structMember = seq(
  optAttributes,
  word.named("name"),
  ":",
  req(typeSpecifier.named("typeRefs"))
).map((r) => {
  return makeElem<StructMemberElem>("member", r, ["name", "typeRefs"]);
});

export const structDecl = seq(
  "struct",
  req(typeNameDecl).named("nameElem"),
  req("{"),
  withSep(",", structMember).named("members"),
  req("}")
).map((r) => {
  const e = makeElem<StructElem>("struct", r, ["members"]);
  const nameElem = r.named.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  r.app.state.push(e);
});

// keywords that can be followed by (), not to be confused with fn calls
const callishKeyword = simpleParser("keyword", (ctx: ParserContext) => {
  const keywords = ["if", "for", "while", "const_assert", "return"];
  const token = ctx.lexer.next();
  const text = token?.text;
  if (text && keywords.includes(text)) {
    return text;
  }
});

export const fnCall = seq(
  word
    .named("name")
    .map((r) => makeElem<CallElem>("call", r, ["name"]))
    .named("calls"), // we collect this in fnDecl, to attach to FnElem
  "("
);

// prettier-ignore
const fnParam = seq(
  optAttributes,
  word,
  opt(seq(":", req(typeSpecifier.named("typeRefs"))))
);

const fnParamList = seq(lParen, withSep(",", fnParam), rParen);

// prettier-ignore
const variableDecl = seq(
  or("const", "var", "let", "override"), 
  word, 
  ":", 
  req(typeSpecifier).named("typeRefs")
);

const block: Parser<any> = seq(
  "{",
  repeat(
    or(
      callishKeyword,
      fnCall,
      fn(() => block),
      variableDecl,
      anyNot("}")
    )
  ),
  req("}")
);

export const fnDecl = seq(
  optAttributes,
  "fn",
  req(fnNameDecl).named("nameElem"),
  // TODO some built in functions can have a template here, e.g. bitcast
  req(fnParamList),
  opt(seq("->", optAttributes, typeSpecifier.named("typeRefs"))),
  req(block)
).map((r) => {
  const e = makeElem<FnElem>("fn", r);
  const nameElem = r.named.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  e.calls = r.named.calls || [];
  e.typeRefs = r.named.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalVar = seq(
  optAttributes,
  or("const", "override", "var"),
  opt(template),
  word.named("name"),
  opt(seq(":", req(typeSpecifier.named("typeRefs")))),
  req(anyThrough(";"))
).map((r) => {
  const e = makeElem<VarElem>("var", r, ["name"]);
  e.typeRefs = r.named.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalAlias = seq(
  "alias",
  req(word.named("name")),
  req("="),
  req(typeSpecifier).named("typeRefs"),
  req(";")
).map((r) => {
  const e = makeElem<AliasElem>("alias", r, ["name", "typeRefs"]);
  r.app.state.push(e);
});

const globalDecl = or(fnDecl, globalVar, globalAlias, structDecl, ";");

const rootDecl = or(globalDirectiveOrAssert, globalDecl, directive, unknown);

const root = preParse(comment, seq(repeat(rootDecl), eof()));

export function parseWgslD(
  src: string,
  srcMap?: SrcMap,
  params: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = root
): AbstractElem[] {
  const lexer = matchingLexer(src, mainTokens);
  const state: AbstractElem[] = [];
  const context: ParseState = { ifStack: [], params };
  const app = {
    context,
    state,
  };
  const init: ParserInit = {
    lexer,
    app,
    srcMap,
    maxParseCount,
  };

  grammar.parse(init);

  return app.state;
}

if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    globalDirectiveOrAssert,
    template,
    typeSpecifier,
    structMember,
    structDecl,
    fnCall,
    fnParam,
    fnParamList,
    block,
    fnDecl,
    globalVar,
    globalAlias,
    globalDecl,
    rootDecl,
    root,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
