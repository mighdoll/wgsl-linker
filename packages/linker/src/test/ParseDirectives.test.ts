import { _withBaseLogger, tokens } from "mini-parse";
import { logCatch, testParse } from "mini-parse/test-util";

import { expect, test } from "vitest";
import {
  ImportElem,
  ModuleElem,
  TemplateElem,
  TreeImportElem,
} from "../AbstractElems.js";
import { argsTokens } from "../MatchWgslD.js";
import {
  directive,
  importing,
  lineCommentOptDirective,
} from "../ParseDirective.js";
import { parseWgslD } from "../ParseWgslD.js";
import { testAppParse } from "./TestUtil.js";
import { last } from "../Util.js";
import { SimpleSegment, treeToString } from "../ImportTree.js";
import { dlog } from "berry-pretty";

test("directive parses #export", () => {
  const { appState } = testAppParse(directive, "#export");
  expect(appState[0].kind).equals("export");
});

test("parse #export", () => {
  const parsed = parseWgslD("#export");
  expect(parsed[0].kind).equals("export");
});

test("parse #import foo", () => {
  const parsed = parseWgslD("#import foo");
  expect(parsed).toMatchSnapshot();
});

test("parse #import foo(a,b) as baz from bar", () => {
  const parsed = parseWgslD("#import foo as baz from bar");
  expect(parsed).toMatchSnapshot();
});

test("lineComment parse // #export ", () => {
  const src = "// #export ";
  const { position, appState: app } = testParse(lineCommentOptDirective, src);
  expect(position).eq(src.length);
  expect(app).toMatchSnapshot();
});

test("lineCommentOptDirective parses #export(foo) with trailing space", () => {
  const src = `// #export (Elem)    `;
  const result = testAppParse(lineCommentOptDirective, src);
  expect(result.appState[0].kind).eq("export");
});

test("parse #export(foo) with trailing space", () => {
  const src = `
    // #export (Elem) 
  `;

  const parsed = parseWgslD(src);
  expect(parsed).toMatchSnapshot();
});

test.skip("importing parses importing bar(A) fog(B)", () => {
  const src = ` importing bar(A), fog(B)`;
  const { parsed } = testAppParse(tokens(argsTokens, importing), src);
  expect(parsed?.tags.importing).toMatchSnapshot();
});

test.skip("parse #export(A, B) importing bar(A)", () => {
  const src = `
    #export(A, B) importing bar(A)
    fn foo(a:A, b:B) { bar(a); }
  `;
  const parsed = parseWgslD(src);
  expect(parsed[0]).toMatchSnapshot();
});

test("#export w/o closing paren", () => {
  const src = `#export (A
    )
    `;
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => parseWgslD(src));
  expect(logged()).toMatchInlineSnapshot(`
    "expected text ')''
    #export (A   Ln 1
              ^"
  `);
});

test("parse #extends", () => {
  const src = `#extends Foo(a,b) as Bar from baz`;
  const appState = parseWgslD(src);
  expect(appState[0]).toMatchInlineSnapshot(`
    {
      "args": [
        "a",
        "b",
      ],
      "as": "Bar",
      "end": 33,
      "from": "baz",
      "kind": "extends",
      "name": "Foo",
      "start": 0,
    }
  `);
});

test("parse extends", () => {
  const src = `extends Foo(a,b) as Bar from baz`;
  const appState = parseWgslD(src);
  expect(appState[0]).toMatchInlineSnapshot(`
    {
      "args": [
        "a",
        "b",
      ],
      "as": "Bar",
      "end": 32,
      "from": "baz",
      "kind": "extends",
      "name": "Foo",
      "start": 0,
    }
  `);
});

test("parse #module foo.bar.ca", () => {
  const src = `#module foo.bar.ca`;
  const appState = parseWgslD(src);
  expect(appState[0].kind).eq("module");
  expect((appState[0] as ModuleElem).name).eq("foo.bar.ca");
});

test("module foo.bar.ca", (ctx) => {
  const appState = parseWgslD(ctx.task.name);
  expect(appState[0].kind).eq("module");
  expect((appState[0] as ModuleElem).name).eq("foo.bar.ca");
});

test("module foo::bar::ba", (ctx) => {
  const appState = parseWgslD(ctx.task.name);
  expect(appState[0].kind).eq("module");
  expect((appState[0] as ModuleElem).name).eq("foo/bar/ba");
});

test("module foo/bar/ba", (ctx) => {
  const appState = parseWgslD(ctx.task.name);
  expect(appState[0].kind).eq("module");
  expect((appState[0] as ModuleElem).name).eq("foo/bar/ba");
});

test("parse import with numeric types", () => {
  const nums = "1u 2.0F 0x010 -7.0 1e7".split(" ");
  const src = `#import foo(${nums.join(",")}) from bar`;
  const appState = parseWgslD(src);

  const segments = (appState[0] as TreeImportElem).imports.segments;
  const lastSegment = last(segments) as SimpleSegment;
  expect(lastSegment.args).deep.eq(nums);
});

test("parse template", () => {
  const src = `#template foo.cz/magic-strings`;
  const appState = parseWgslD(src);
  expect((appState[0] as TemplateElem).name).deep.eq("foo.cz/magic-strings");
});

test("#import foo from ./util", (ctx) => {
  const appState = parseWgslD(ctx.task.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).eq("./util/foo");
});

test('import { foo } from "./bar"', (ctx) => {
  const appState = parseWgslD(ctx.task.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).eq("./bar/foo");
});

test('import { foo, bar } from "./bar"', (ctx) => {
  const appState = parseWgslD(ctx.task.name);
  const imports = appState.filter((e) => e.kind === "treeImport");
  const segments = imports.map((i) => treeToString(i.imports));
  expect(segments).includes("./bar/foo");
  expect(segments).includes("./bar/bar");
});
