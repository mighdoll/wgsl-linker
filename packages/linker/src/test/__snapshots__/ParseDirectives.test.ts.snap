// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importing parses importing bar(A) fog(B) 1`] = `
[
  [
    {
      "args": [
        "A",
      ],
      "end": 25,
      "kind": "import",
      "name": "bar",
      "start": 11,
    },
    {
      "args": [
        "B",
      ],
      "end": 25,
      "kind": "import",
      "name": "fog",
      "start": 11,
    },
  ],
]
`;

exports[`lineComment parse // #export  1`] = `
[
  {
    "end": 11,
    "kind": "export",
    "start": 3,
  },
]
`;

exports[`parse #export(A, B) importing bar(A) 1`] = `
{
  "args": [
    "A",
    "B",
  ],
  "end": 36,
  "importing": [
    {
      "args": [
        "A",
      ],
      "end": 35,
      "kind": "import",
      "name": "bar",
      "start": 29,
    },
  ],
  "kind": "export",
  "start": 5,
}
`;

exports[`parse #export(foo) with trailing space 1`] = `
[
  {
    "args": [
      "Elem",
    ],
    "end": 24,
    "kind": "export",
    "start": 8,
  },
]
`;

exports[`parse #import foo 1`] = `
[
  {
    "end": 11,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "foo",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;

exports[`parse #import foo(a,b) as baz from bar 1`] = `
[
  {
    "end": 27,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "bar",
        },
        SimpleSegment {
          "args": undefined,
          "as": "baz",
          "name": "foo",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;
